Design doc: direct artifacts upload (#13134)
readthedocs/readthedocs.org: docs/dev/design/direct-artifacts-upload.rst

Build anywhere, host with the best
==================================

Goals
-----

- Build the documentation anywhere, and upload the artifacts to Read the Docs.
  Zip it, ship it, host it.
- Integrate with existing modeling as much as possible.
- Easy to transition/test from our current build system for existing projects.

Non-goals
---------

- Complete re-design of the build/version system/modeling.
- Deprecate our current build system for existing projects.

>   We want to support both for a while, and let users transition at their own pace.

Reviewer: ```suggestion
  We want to support both for the long term. 
```

Motivation
----------

- Most requested feature since ever.
- Security: we would no longer run arbitrary code on our servers.
- Performance: some projects have very long build times, our builders are slow compared to other CI services.
- Flexibility: we aren't as flexible as other CI services. We aren't a CI service.
- Easy to integrate: users already build their docs in their CI, no need to replicate that on our side.
- Best of both worlds: users can build their docs anywhere as they want, but still get the benefits of hosting on Read the Docs.

Proposed architecture
---------------------

This is a high level overview of the proposed architecture for the new build system.
The details about each phase of the process are explained in the following sections.

- Users build their documentation artifacts (HTML, PDF, ePub, etc.) in their CI or locally.

> - Users call a script or GH action to upload the artifacts to Read the Docs.

Reviewer 2: > Users call a script or GH action

I would leave this for the v2. We should focus on building the infrastructure and API first.

I would focus on being able to do a simple cURL command first. I also don't want to force users to deal with "installing a script" or similar. If we use Python, they will need to know about `pip` or `uv`, etc. If we provide a simple cURL command they don't need anything else.

My priorities about this would be: cURL, GH action, script -- in that order.

Author: Using curl is basically the script, we shouldn't require users to install python to run our script, using bash should be fine.

Reviewer: Yea, I think a basic script is quite simple.

Reviewer 2: > basic script is quite simple

Bash scripts are never simple 😄 -- in particular when you start needing to check for variables, and command outputs 🙃 .

Anyways, if we have a 5 cURL commands script, that's fine to me 👍🏼 ; but I doubt we will end up with something that simple.

Reviewer 3: Yeah it sounds like what we want is not going to be simple. I'd rather we wait and give the user something a bit polished. We're asking them to install something locally and in GH actions, etc and we don't want to have to support something half baked or worry about backwards compatibility.

It sounds like we want more out of a script anyways -- checks, validation, uploading.

Author: The final scrip won't be short, but it would still be simple for most cases. Trying to support different OS, specially windows will make things complicated for sure.

Reviewer 3: You'll hit cross OS support with a shell script even with MacOS and Linux. We don't need any scripts to start testing this feature out though, so I still think the correct call for a first iteration is to skip writing anything.

My issue with bash scripts besides distribution is that I see this script gaining features and complexity and I don't think we have a strong skill set in maintaining shell scripts.

>   - A token is previously generated on Read the Docs and provided to the user to interact with the upload API (internal API).

Reviewer 2: It would be good to implement the token to be scoped to that particular build only, so we can re-use this endpoint for "Isolated builders" as well: https://github.com/readthedocs/readthedocs.org/pull/13133#discussion_r3417787660

Author: The upload token needs to be scoped to the project, since it is used to create the  builds.

  - Some pre-checks can be done on the client side before uploading, like checking the size of the artifacts, validating the metadata, etc.
    This is only to avoid unnecessary uploads and provide faster feedback to users.
    All the checks will also be done on the server side.
  - All artifacts are zipped into a single file.

>   - A post request is made to the upload API with some metadata like Git branch/tag, commit hash, configuration, etc.

Reviewer 2: This step should be optional --or at least, most of this data should be optional. Users may not be using Git, for example.

Author: I think it should be required, but since we aren't validating that commits/branches exist, it can be anything to identify this version/build. I'd be okay with leaving the commit out of it, but maybe in the future.

Reviewer: We should try to keep it as similar as possible to our existing flow, so we don't break the UI. We can always relax the constraints later.

Reviewer 3: Yeah, plus we provide things like authentication against the repositories. I think the expectation should still be that users always use a Git repo for their project regardless of whether they are manually pushing files or not.

  - A build object is created in Read the Docs with the status "pending" and a presigned URL is returned to the user to upload the artifacts.

>   - The zip file is uploaded to the presigned URL to a temporary location in S3.

Reviewer 2: ```suggestion
  - The user uploads the zip file to the presigned URL to a temporary location in S3.
```

>   - A post request is made to the upload API to notify that the upload is complete and the build can be processed.

Reviewer: Can we not get this from S3? Maybe in v2?

Reviewer 2: @ericholscher do you mean like a subscribe to a S3 event or receive a webhook from it?

Reviewer 2: @stsewd is this possible?

Author: https://github.com/readthedocs/readthedocs.org/pull/13134#discussion_r3455374092

- Read the Docs triggers a task to process the build.
  - A docker container is created to process the build (like our current builders do).
  - The zip file is downloaded from the presigned URL.
  - The artifacts are extracted and validated.
  - The artifacts are uploaded to their final destination and pre-processed for indexing (same process as our current builders).
  - The build status is updated to "success" or "failure" depending on the result.
- The documentation is available to be served to users.

Versions
--------

Versions are created when the upload is initiated and linked to a Git branch/tag/commit, just like our current versions.
We don't enforce that the branch/tag/commit exists in the Git repository, we just assume it does.
This is useful to keep the same modeling for versions, and most of the UI the same.

> If the version for that Git branch/tag/commit already exists, it will be reused for the new build.

Reviewer: This seems a bit scary.

Author: Which part? re-using versions is what we already do, like, we don't create a new version for every build.

Reviewer 2: What's the reason to allow overriding the current artifact for an existing version with a one built from the upload? If we really want that, I would add a new field `Version.allow_upload_override(default=False)` to allow this and avoid users shooting on their own foot.

Author: We already override artifacts on each build. Builds from branches need to be overridden. See https://github.com/readthedocs/readthedocs.org/pull/13134#discussion_r3456819377 for something similar.

Reviewer: I think in particular:

```
We don't enforce that the branch/tag/commit exists in the Git repository, we just assume it does.
```

This would lead to Versions getting detached from the Git state, which could be pretty messy?

Author: But we don't have the need to enforce this. We use commits to:

- links
- build status

Don't think we have the need for validation for that.

Reviewer: I'm not worried about commits, but about versions. If we have a git repo with a version that doesn't exist, that is going to make it really hard to resync with a git repo for new/updated versions, right? I guess if it's uploaded it will already be excluded and labeled in the UI somehow?

Reviewer 2: This is not clear to me. @ericholscher are you suggesting we shouldn't allow users to upload artifacts for versions that don't have a corresponding branch/tag on Git?

Author: For me this is the same as having a branch deleted that maps to an active version. When that happens the version isn't deleted on RTD, if the branch is re-created, nothing will happen, our system will map that version to the branch again.

Reviewer: Yea, I'm trying to keep the mapping to git repos somewhat defined. 

It's a good point that an active version that has its vcs branch deleted will still exist, so it's in a similar state I guess. 

>  if the branch is re-created, nothing will happen, our system will map that version to the branch again.

The main concern I have is the user saying "I want to resync the versions with Git" -- and what do we do in this case? It will be a VCS-backed version, but it will have been uploaded from some unknown version? So the VCS <-> output mapping gets broken... but maybe that's just unavoidable?

Reviewer 2: > The main concern I have is the user saying "I want to resync the versions with Git" -- and what do we do in this case? It will be a VCS-backed version, but it will have been uploaded from some unknown version? So the VCS <-> output mapping gets broken... but maybe that's just unavoidable?

This is exactly what I want to manage by having a specific field in the `Version` object: with https://github.com/readthedocs/readthedocs.org/pull/13134#discussion_r3473495152

When the versions are re-synced from Git, if the branch/tag (re)appears in Git the behavior will be defined by `Version.allow_upload_override`:
- `True`: no builds are triggered from the tag/branch
- `False`: the version is (re)built from tag/branch

That's what I meant when I said I want to protect users from shooting themselves in the foot. Being explicit about what type of version is, will avoid unexpected behavior. We will have two types of versions:

- Version built from Git
- Version built externally and uploaded via the API

The user can switch manually between them whenever they want and change this behavior.

Versions will have a new field to indicate the source of their latest build (``build_source``), either "rtd-build" or "external-upload".
This will allow us to do some checks and validations.

Versions marked as active and where their latest build is from an external upload can't be built using our current build system.
This is to avoid overwriting the documentation built from an external upload with a build from our current build system.

> This should mainly happen while users migrate to the new system, new users should only use the new upload system.

Reviewer: How do users reset to our build system?

Author: We can manually change the attribute of the version. But I think when we ship this as a beta feature, we should expose an option in the project, so it always prefers builds from the upload API only or from our builders. Mixing them would be confusing.

Reviewer: You mean you project would be configured so only 1 type of builds would be possible at a time? I imagine people will hit edge cases with that pretty quickly (eg. I want RTD to build most of my versions, but I need to upload these old versions periodically).

Author: One type of build at a time is less confusing. I think the opposite will happen, users will favor using the GH action for everything.

Reviewer 2: Yeah, I agree with @ericholscher here. I'm sure I want this setting to be per-version (https://github.com/readthedocs/readthedocs.org/pull/13134#discussion_r3458544944) instead.

I want _"some versions to be built with my current system (Read the Docs) and I will start building the new ones using the new feature"_.

> eg. I want RTD to build most of my versions, but I need to upload these old versions periodically

This is one of the most requested features as well that I want to support.

Author: It's already per version, but users don't have control over it. It's only used to prevent accidental overrides while using both systems.

Reviewer 2: Well, yeah, I'm saying that _as a user_ I want to control this by myself. I mentioned this earlier, I prefer having a field `Version.allow_upload_override` that's `True` for versions created via the upload API and `False` for versions _we_ create when syncing repositories.

Then, if _the user_ wants to change a version to `upload_override=False` to upload docs via the API, they can change it under the Edit version page:

<img width="565" height="630" alt="Screenshot_2026-06-25_11-57-59" src="https://github.com/user-attachments/assets/2ba08d99-adb6-4069-894d-a7459456a254" />

Reviewer: Sure, that's one use case, but users who want our build system but also want to upload some old versions seems like a very obvious use case as well? This feels like an artificial limitation? 

I feel like this is required because of the proposal to allow versions to be detached from git. I think we likely want something closer to versions still being tied to Git, and allowing users to upload specific versions of known Git Verisons?

Reviewer 2: I'm not adding any limitation with what I'm saying here.

If the user hits the API for a version that it's not on Git, we will create it with `allow_upload=True` and cannot be changed to `False` unless the user creates a branch/tag with the same name on Git repository.

If you want to upload an old version you don't have on Git, just hit the API with `{"version": "pretty-old"}` and we will create that version for you, unzip the file and serve the docs.

Reviewer 3: Do we need to decide on this to continue? It feels like we might need to pick something here to continue.

I think I agree with what @humitos is proposing. The framing could maybe be different than "allow upload" specifically, but I think giving the user and option to select on the version makes sense. This feels maybe related to the `machine` field in that we already have a representation of whether the version is an actual Git construct or not.

The API interaction to create a version disconnected from Git feels hard for users to use up front, but it could be part of whatever script we create eventually and mostly hidden from the user. I think the UX of requiring a Git branch would be odd, we'd have to give an error to the user that tells them to first create an unused Git branch and then try uploading again, which seems confusing.

Reviewer 2: Yeah, I agree with @agjohnson here 👍🏼 

My thoughts:

- I wouldn't require a Git branch/tag to upload a _new_ version
- We should have 2 different type of versions: allow upload, and don't allow upload (this is the new field I'm proposing)
- Git re-sync behavior will be determined by this new field (see https://github.com/readthedocs/readthedocs.org/pull/13134#discussion_r3504566906)
- We can name the field as we want. I don't have strong opinions there, but it should clearly communicate that particular `Version` can be overwritten via the upload API
- Versions that are not allowed to be overwritten via upload API should be "protected" and fail with a clear error

Author: > Do we need to decide on this to continue? It feels like we might need to pick something here to continue.

I don't think we need do decide this now. The document already describes a mechanism to protect users from mistakes, exposing it to users could be a next step if needed.

> This should mainly happen while users migrate to the new system, new users should only use the new upload system.

Reviewer 2: > new users should only use the new upload system

Why? We are not planning on deprecate the current build system.

Author: No, but users shouldn't mix both ways to build/upload files. That's very error-prone.

Reviewer 2: OK, we need to adapt this sentence accordingly.

We are having the same conversation at https://github.com/readthedocs/readthedocs.org/pull/13134#discussion_r3455315519. Once we decide how to proceed there, we should come back here and update the text.

Author: Feels like the other discussion is different. On a new project, if you start with uploading files directly, what's the use case to go back to use our builders?

Reviewer 3: There are many use cases where we'll see projects with mixed versions, I don't think we should think of this as two exclusive build systems. Projects that needs to update old versions but want to continue building with us is a perfect example of this -- new builds are still our standard builds and their historical builds are uploaded.

Very big 👍 on not framing this as a replacement for our build system either way. It's just an alternative and not the main recommended path.

Reviewer: I feel like this is still the biggest open question for me. Are we going to allow both systems at the same time? How will that get expressed in the UI? It doesn't seem impossible, but we definitely want to make sure the interactions aren't super weird.

Author: We can support both, but yeah, the interaction and UI is going to be a little confusing. For me, new projects won't have the need for using our builders if they start with DAU (direct artifacts upload, sorry, another acronym :D). Existing projects will:

- Start using DAU, and not require builders
- Keep using our builders, but require DAU for one-shot occasions.

Reviewer 2: > For me, new projects won't have the need for using our builders if they start with DAU

This _"could be true"_ eventually, but we are pretty far away from there.

All the documentation, and all the projects, on the internet are using the our builders. Every single example. Almost every new project will keep using our builders for a long time until the community can spread the word about the new DAU system and they can prove it's better/simpler than the older one.

During that time, we need to support both at the same time for the same project and make it nicely work on the UI, providing a good UX.

We will no longer update the version to keep in sync with the Git branch/tag/commit,
this will be updated after the upload is complete and the build is processed for that version.

> Active versions should reflect the state from where the documentation was built.

Reviewer: Not sure what this means?

Author: Like, we shouldn't change the commit of a version without having a successful build for it. But I guess we would have the same pattern here for now.

Versions that are deleted from the Git repository will be deleted from Read the Docs if an automation rule is configured to do so.
This includes versions created from external uploads, as they are still linked to a Git branch/tag/commit.

.. note::

   With this new system, we don't have the need to keep a copy of the Git repository versions in Read the Docs.
   We can discuss in the future how we want to handle that.
   The only reason to keep it is to be able to trigger builds from our current build system (while projects migrate),
   and to keep automation rules working on deleted versions.

The "build version" option should be disabled/hidden for versions created from external uploads.

Builds
------

> Builds will have a new field to indicate the source of the build (``build_source``), either "rtd-build" or "external-upload".

Reviewer 2: I would simplify the value here.

```suggestion
Builds will have a new field to indicate the source of the build (``build_source``), either "internal" or "external".
```

Author: We already use internal/external for PR builds

Reviewer 2: Yeah, that's fine, those are for `Version`. We are in a different context here: `Build`. The Version could be internal and the Build external; and all the other combination make sense as well.

Author: the build model has also the external/internal concept for builds from PRs

Reviewer 3: Do we expect a third value? Why not just a boolean `is_uploaded`?

Author: I'm fine with that too

This will allow us to differentiate between builds that were triggered by our current build system and builds that were triggered by the new upload system.

Latest triggered build wins, previous build is cancelled. Just like our current build system.

Statuses:

> - A build is created when the upload is initiated, with the state "uploading".

Reviewer: I don't understand why we have 2 different names for these states? We should just use the existing state and order that the state happens on an RTD build to keep things simple. (Basically the "reused" states).

Author: We do have some other things/tasks linked to those states. But I'm okay with re-using the states for now, but I do think is more clear for the states to match the actual state of the build.

Reviewer 2: Yeah, I would re-use the same states for now. Otherwise, it would be pretty confusing.

We can add notifications if we want to communicate something extra, like "Waiting for the user to upload the artifacts..." immediately after the build is created.

  The "triggered"/"building" state could be re-used.
- When the upload is complete, the state is "uploaded" (can also be "in queue" or "ready to process").
  The "triggered"/"building" state could be re-used for this.
- When the build is being processed, the state is "processing".
  The "building" state could be re-used for this.
- When the assets are validated and uploaded to their final destination, the state is "finished".
- The success attribute is updated to "true" or "false" depending on the result of the processing.
- If the build is cancelled, the state is "cancelled".
  Builds are cancelled when a new build is triggered for the same version,
  or when the user cancels the build manually.
  Just like our current build system,

> The "rebuild" option should be disabled/hidden for builds created from external uploads.

Reviewer: We probably need a way to handle _our_ stuff breaking in these cases. Maybe just with retries on our side?

Author: Yeah, but we can't have a re-build over very old builds, that would mean saving the uploaded artifacts for a very long time.

Reviewer 2: Yeah, I agree with both here 😄 

There is no reason to tell the user to re-upload if "our search index failed", for example. However, to re-trigger the search index, we need to have the docs to index.

Author: Our indexing tasks can be re-tried, but we can't re-try unzipping.

Reviewer 2: > but we can't re-try unzipping

Why not? We have the .zip in S3. We just need to trigger a new build for it.

Author: They can be re-tried, but only immediately, not after a day.

Reviewer 2: Yeah, that's fine. We can increase the time we keep the zip file on S3 if we need more time than that

Author: Not sure if we need to expose that retry operation to users. Our app should retry immediately.

Upload API
----------

The upload API will be a private API, and will require an "upload token" to authenticate the requests.

> This token can be generated by users in their project settings.

Reviewer 2: I think we don't want the upload token to be tied to a user, but to a project instead. This is something users has been asking for a while now. They don't want to put their own token in the CI where everybody in the team has access to.

Author: Yeah, the token is linked to a project. This is describing how to get it, a user needs to go to the project settings to get it.

Reviewer 2: Ah, you are right. I miss read it.

.. note::

>    The user token can be re-used for initial testing.

Reviewer: 👍

> POST /api/v3/_internal/upload/initiate:

Reviewer: I really don't like the `_internal` API concept. Why is that needed? Shouldn't it just be `/api/v3/upload/initiate/`?

Author: To make it extra explicit that this isn't a user facing API, I'm okay with having it undocumented only.

Reviewer 2: This is not an internal API endpoint. This is how _users_ will trigger a new build in this new system.

Author: What is public is the script to upload artifacts, the APIs should be private. Users shouldn't hit the APIs by themselves.

Reviewer: I think a script that users are managing is a public API. I don't know how we would stop them from calling it, and I don't think it's a private or internal API in any meaningful sense. It's definitely public and meant to be used by users.

Author: We shouldn't expect or encourage users to use the API directly, only via the script.

Reviewer 2: > We shouldn't expect or encourage users to use the API directly, only via the script.

I don't agree with this sentence 😄

Author: No other service gives direct access to their API, only through SDKs/clis/scripts.

Reviewer 3: This has to be public for users to use it either way, obfuscating the path doesn't really do much to discourage direct use. I agree we just follow the URL structure we're already using. We don't have to document it publicly if we think that's necessary.

Reviewer: Yea, public but undocumented is the right balance if we don't want people to use it.

  Initiate the upload process and create a build object with the state "uploading".
  This endpoint will receive some metadata like Git branch/tag/commit, configuration, etc,
  and return build object and the presigned URL to upload the artifacts.

  If we don't find a version for the Git branch/tag, we can create a new version for it,
  and attach the build to that version.

  ::
      {
         "name": "main",  // name of the Git branch/tag or PR number.
         "type": "branch",  // or "tag" or "external" (PR).

>          "commit": "abc123", // full commit hash

Reviewer 2: These should be optional too. There will be users that don't even use Git.

Author: I'd leave that use case for later.

Author: We already support git only

Reviewer: Yea, I'm fine requiring git-like concepts to start. We can always use dummy data for this in the client as needed (eg. unix timestamp)

         # Optional, if not provided, it's derived from the name of the branch/tag.
         # If the version with that slug exists, but it points to a different Git ref, we error out.
         "slug": "main",
         "configuration": {
            "search": {
            }

>          }

Reviewer 2: 👍🏼 -- This should have the same structure of our current YAML config for the supported options. We shouldn't create a new format here.

      }

   The presigned URL will have a limit on the size, content type, and expiration time.
   See https://docs.aws.amazon.com/boto3/latest/reference/services/s3/client/generate_presigned_post.html.

   The zip to upload should contain all artifacts in the following structure:

   - html/index.html
   - pdf/documentation.pdf
   - epub/documentation.epub
   - zip/documentation.zip

>    The names of pdf/epub/zip files can be anything, as long as there is just one.

Reviewer: We should use existing paths for this that are namespaced: https://docs.readthedocs.com/platform/stable/build-customization.html#where-to-put-files

Author: This is the structure we control after we create the zip, the only mismatch is zip -> htmlzip, I can update the document to reflect that.

Reviewer 2: yeah, I agree with @ericholscher, we should use `htmlzip/` here as well.

POST /api/v3/_internal/upload/complete:
  Notify that the upload is complete and the build can be processed.

>   This endpoint will receive the build ID.

Reviewer: We should just make the input/output json explicit in this doc -- it will be easier than the text to understand. 

But I do wonder if we even need this? I would expect us to handle this ourselves based on the file upload completing.

Author: Added the input and output.

  This endpoint will trigger a task similar to our builders.
  This task will start a docker container (we can re-use the docker image used to clone the repository),
  download the artifacts, validate them, upload them to their final destination, and update the build status,
  send the webhooks, update the commit status, etc.
  Just like our current builders do.

  The zip validation includes checking that the expected files are present,
  and that the size of the artifacts is within the limits.

  This endpoint could also be used to notify that the upload has failed,

>   so the build is cancelled and the status is updated to "failure".

Reviewer: Not sure what this means? Like, if they hit this endpoint, we check if the file exists, and if not, error? Seems like this should be part of the upload process.

Author: After the upload has failed, we don't know about it (even after retry), the /complete endpoint will let us know about it instead of having to wait for the build to expire.

Reviewer: Not sure I follow? Who is `we` and what is `it`? You mean the `rtd-upload` script doesn't know about the state of the upload?

Author: Our application doesn't know if the upload failed or not, in order to know about it the task needs to retire that build. A faster process is for the script to communicate that using the API.

Reviewer: Right.. but that's an error case? Why should that second API call be required when everything works properly?

Reviewer 2: I understad that the `/complete` endpoint is used to _communicate our app that the upload is finished and the build itself has to be started (unzipping, etc)_. The app _needs_ a way to know this somehow. If it's not manually via an endpoint, it should be an S3 webhook or similar --but someone needs to communicate this to the app.

Author: > Right.. but that's an error case? Why should that second API call be required when everything works properly?

I think this goes back to the discussion on using S3 events vs the API https://github.com/readthedocs/readthedocs.org/pull/13134#discussion_r3455374092

  We also need to have a task to cancel builds that are in "uploading" state for a long time,

>   in case the upload is never completed.

Reviewer: We should specify the limits here. 4 hours?

Author: Could be even less. If the presigned S3 URL expires after 30 min (could be 15 min, maybe 30 for slow connections?), the builds can be marked as terminated after 45 min.

Reviewer: That assumes our task queues aren't backed up at all. I don't think that's a good assumption.

Author: This is for builds that have started the upload process only.

  .. note::

     S3 sends events when a file is uploaded, while we can use that to trigger the processing task,

>      it's better to have an explicit endpoint, it lets us report failures easily.

Reviewer: Should have put this higher up :) 

Presumably we would report failures as an outgoing webhook, or a upload failure? It feels weird that it would be a different process than the upload? What failures are we worried about other than "upload not complete" ?

Author: Feels like a better UX in general, builds are marked as failed immediately, instead of waiting for our task to terminate them in case the upload failed. S3 doesn't have a way to know that the upload failed, it only knows if the artifacts were uploaded. For development is also easier, we don't depend on having an AWS account to test things (but I do think we should test more with a real AWS account). We also would need to have shared secret in AWS instead of a per-project/user token. So, in general I prefer having an explicit enpoint.

Reviewer: But the client knows if the upload has failed -- isn't that where the error report will be presented? 

The dev argument and the "the server doesn't know an upload failed" here are reasonable ones, but I'd like to understand a bit more exactly what we're worried about around production. We're still depending on the client to properly implement an API contract, so the client could just have a way to cancel the build on failed upload as an exception handler or something, instead of always requiring a second API call.

Author: >  We're still depending on the client to properly implement an API contract, so the client could just have a way to cancel the build on failed upload as an exception handler or something, instead of always requiring a second API call.

That's why we should be the ones providing the script to zip and upload files instead of having users manually call the APIs.

Reviewer 2: > We're still depending on the client to properly implement an API contract, so the client could just have a way to cancel the build on failed upload as an exception handler or something

I'm assuming the client will call our current APIv3 to cancel the build: `POST /api/v3/projects/(str: project_slug)/builds/(int: build_id)/cancel/`. This endpoint doesn't exist yet, but it make sense to add it.

Reviewer: I feel like you're leaning really heavily on the API being only used by us, but that's not a realistic assumption. Users will definitely want and need to customize what is happening on the client, and we shouldn't design a public API in a way that is only meant to be used by us. I don't think it's a real constraint or expectation -- people will definitely use it if we expose it. 

Unless there is a physical limitation on the usage, we should assume they are always public.

Author: What are the use cases that a script won't support? No other services allow direct API access. The script/GH action will allow for overriding any parameters, users don't need to interact with the API or know about the flow that's used to upload the files.

Reviewer: People will definitely want to do weird stuff (eg. move files around based on the state of the zip file or something). I guarantee it. But I don't think it really matters, we just can't assume our client is the only thing that will ever use these APIs unless we make them authed or something.

Author: I don't think we should design things around edge cases. No other services allow direct API access. Moving files is already supported, generating the zip happens over the already generated files.

.. note::

>    `OIDC <https://docs.github.com/en/actions/concepts/security/openid-connect>`__ can be explored in the future for tokenless authentication.

Reviewer: Yea, something like Trusted Publishers on PyPI?

Author: yep

S3
--

All artifacts will be uploaded to S3, and then processed from there.
We'll have a separate bucket for the uploaded artifacts,
with a lifecycle policy to clean up the artifacts after a certain period of time.
See https://docs.aws.amazon.com/AmazonS3/latest/userguide/lifecycle-expire-general-considerations.html,
https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-lifecycle-mgmt.html.

Artifacts will be uploaded as a zip file, in the path: {project_id}/{build_id}/artifacts.zip.
Once the build is processed, the artifacts will be extracted and uploaded to their final destination as we do now.

Upload script / GitHub Action
-----------------------------

In theory, given the token, users can upload the artifacts from anywhere,
but we shouldn't encourage users to manually call the API directly.
We can provide a simple bash script that can be downloaded and used to upload the artifacts,
or a GitHub Action that can be used in their workflows.

> 

Reviewer 2: I mentioned this earlier also. I'm not opposed to give users a helper (or something that's pre-baked) and easy to use for them. However, I want to allow users to use regular cURL as well -- that should be the simplest way to use this feature.

Basically, 3 cURL command should be enough to use this feature:
1. API: trigger the build and get the token
2. S3: upload the artifacts
3. API: finish the build

We can add as much as we want around that with the script/gh action; but we shouldn't force users to use them.

Author: Using the script should be the easiest. Otherwise, users would need to extract the parameters for the API call, prepare the zip, and then get the response to do the upload, and manage errors. Any change to the API would break their integration.

Reviewer 3: I don't see much value in doing this manually vs using a script, but I also don't see much value in protecting this API. What are we actually worried about here?

I think it's likely we'll give a mess of shell commands out initially, but longer term we're aiming for a fairly through pass at an installable script. I'd almost rather skip distributing a copy paste version of a bash script if we know we want an installable package that is pretty complex.

Author: I wouldn't distribute this as a copy/paste script, but as a script you download from us or is distributed with the GH action.

Reviewer 3: Downloading a script from us is effectively the same thing as copy pasting a script though. It's not the delivery mechanism that's the problem, it's that the script can't be easily upgraded and so it won't be up to date.

Package managers already have distribution and versioning figured out and users know how to use them. We should plan on distributing through PyPI probably, not reinventing a wheel.

Reviewer: Agreed. We can distribute a bash script via pypi if needed even (though should just write it in Python at that point :D), and easily get the semantics we want. It will plug into our existing requirements management infra as well, which is nice 💯 

GH actions also has a crappy package management builtin, but presumably we'd use that too.

Reviewer 2: I'm a little worry about using PyPI for distribution. This will be tied to the Python community. I'm sure that people building with any non-Python framework will know nothing about `pip`, `uv` or similar and the barrier will be higher for them.

Author: > Downloading a script from us is effectively the same thing as copy pasting a script though. It's not the delivery mechanism that's the problem, it's that the script can't be easily upgraded and so it won't be up to date.

Uploads will start by downloading the latest version of the script, it's the same as installing from pypi, isn't meant for the first download to be re-used.

Reviewer: > Uploads will start by downloading the latest version of the script,

Where is this specified? What this is saying isn't clear to me from the design doc. But it sounds like "what should the client be responsible for" is the other big outstanding design doc.

Author: Added a paragraph to make it more explicit.

Reviewer 3: > Uploads will start by downloading the latest version of the script, it's the same as installing from pypi, isn't meant for the first download to be re-used.

I'm confused by this. What is orchestrating the initial upload if not the script? If I have a docs in repo locally, how do I run the script that doesn't exist on my system yet?

or fallback to read the configuration file from the repository if it exists.

When the upload is initiated, we can pass the configuration to the API,
which will be stored in the build object and used when processing the build.
We already store the configuration in the build object.

.. note::

   Currently, only the search configuration is relevant for processing the artifacts.
   It can be left out for a future iteration if we want to simplify the initial implementation.

Aliases: latest/stable
----------------------

After a build has been processed, we can decide if latest or stable aliases should point to the results of that build.
If so, we trigger a task to update the latest/stable.

This task will only do a copy of the artifacts from the version to the latest/stable location.
While this task won't spin up a docker container, it will create a build object with some dummy output
to keep the same modeling as our current builders, and to be able to send the same webhooks.

.. note::

   We can start thinking about making latest/stable real aliases instead of copying the artifacts to a new location,

>    or duplicating builds for the same source.

Reviewer: Bring back symlinks! 😆

Author: Storing artifacts per-build allows us to do this, and also solve the problem with FTD, and atomic builds :)

Going to create an issue with this idea or comment on an existing one.

Reviewer: Yea, I ran into this a bunch with the visual diff work. I think it's definitely something to consider more.

Author: Left some ideas in https://github.com/readthedocs/readthedocs.org/issues/13044#issuecomment-4782434253

As we allow right now, users can also manually create versions named "latest" or "stable" to skip our logic.

Pull requests

> -------------

Reviewer: This permissions issue seems like quite a large UX issue. I feel like there isn't really a solution presented here?

Reviewer 2: > We can try to offer a token that's only scoped for PR builds.
> This token should still be kept secret, as it can be used to upload builds to the pull request previews domain of the project,
> but the impact of the token being leaked is much smaller, as it doesn't affect production documentation.

Isn't this the proposed solution? 🤔

Author: Is not a great solution, but not sure if we have other way around it. I looked at how codecov manages this, and they offer a "no-token" option, that allows anyone to upload coverage files, which seems terrible, not sure how they handle having coverage files for the same commit/PR.

Reviewer: Yea, that solution still seems bad if it's meant to be kept secret, it seems like it isn't really solving the issue. 

If there isn't a good, secure way to make PR builds work, that's a huge downside.

Reviewer 2: > If there isn't a good, secure way to make PR builds work, that's a huge downside.

Yeah. We should invest more time thinking about a solution here.

Reviewer 3: Our stance could simply be the pull request builds are not supported by uploaded HTML. If you want PR builds, you must use our platform.

Reviewer 2: I understand we are talking about that only "external pull requests" are not supported, right? I mean, pull requests opened by external contributors, but pull requests opened by users with access to the repository will still work for artifact uploads, right?

If that's the case, I'm happy to leave that use case for later and continue researching about it. It's a downside, but not as huge as saying "PR doesn't work with artifact uploads".

Reviewer: Yea, private repos for example are totally unaffected here. I think GH has support for "only use secrets when the PR is opened by a collaboator" or similar, which is likely what we need.

Author: yeah, PRs from branches of the same repo work fine

Reviewer 2: OK. Yeah, I think that's a limitation we could assume for now.

But for PRs that originate from forks, it's more complicated, as the user doesn't have access to the repository and can't generate a token to upload the artifacts,
there are ways to share this token using a ``pull_request_target`` event in GitHub Actions, but it can be easily misused to run malicious code in the context of the base repository.

We can try to offer a token that's only scoped for PR builds.
This token should still be kept secret, as it can be used to upload builds to the pull request previews domain of the project,
but the impact of the token being leaked is much smaller, as it doesn't affect production documentation.

Webhooks and status API
-----------------------

Webhook events:

- Build triggered: we can emit this one when the upload is initiated,
  or replace it with "uploading".
- Build passed: we can emit this one when the build is finished and successful.
  Same as our current "build passed" event.
- Build failed: we can emit this one when the build is finished and failed.
  Same as our current "build failed" event.

Commit status doesn't require any changes, we can update the status as we do now with the three states: pending, success, failure.

> 

Reviewer 2: I would leave this out of the scope of this feature for now and consider these webhooks later together with regular builds as well.

First iteration
---------------

The first iteration would be like a half polished proof of concept, to validate the architecture and the implementation.
While not all things discussed in this document will be implemented in the first iteration,
this first iteration can be improved in the following weeks without major changes.

- Implement the upload API and the processing task.
- Internal bash script to upload the artifacts using the API.

>   Almost not client side validations.

Reviewer: ?

Reviewer 2: Let's just use cURL commands for the first implementation.

Author: Client side validations are checking that the correct files are going to be uploaded to give a faster feedback to the user. I'm proposing skipping those checks and just upload things to keep the upload script small.

Reviewer: Gotcha. I think the typo tripped me up. 

```suggestion
  Almost no client side validations.
```

- Skip search configuration.
- Add the new S3 bucket with the lifecycle policy.
- Skip latest/stable logic.
- Re-use the user token for authentication.
- Minimal integration in the UI.
- Minimal checks/enforcement for versions and builds created from external uploads.

> This first iteration is not a beta, it will be used for our own projects only.

Reviewer: This seems like a great initial proof of concept.

Future improvements and ideas
-----------------------------

- Use a more lightweight docker image to unzip the artifacts.
- Use a lambda instead of a docker container to process the build.
  Lambdas have a maximum storage of 10GB, which may not be enough for some builds.
  But AWS has some other options for big files:

  - https://aws.amazon.com/blogs/storage/automated-extraction-of-compressed-files-on-amazon-s3-using-aws-batch-and-amazon-ecs/
  - https://aws.amazon.com/blogs/storage/automatically-decompress-files-in-amazon-s3-using-aws-step-functions/

> 

Reviewer 2: Based on our discussion about architecture for isolated builders, I wouldn't change the process here and re-use as much as possible, instead of having two different patterns.

Author: The only process builders will be doing is unzipping the files, we don't need a whole builder for that.

Reviewer 2: Yeah, but the architecture complexity created "just to unzip a file" is too big compared to re-using our current build process. We just need to short-circuit the process and instead of build the docs, unzip the file.

Author: I'd say it's the opposite. Using our current build system just to unzip a file adds a lot of complexity. But this is left for a future improvement.

Reviewer 3: The complexity comes from added operations and management overhead, not code. We all know and can debug the current build system, we have very little shared knowledge of lambdas.

Reviewer 2: > Using our current build system just to unzip a file adds a lot of complexity

I don't agree, here. It should be a lot simpler. Look at what @ericholscher implemented as a simplified version of what I'm saying at:

- https://github.com/readthedocs/readthedocs.org/pull/13011/changes#diff-2c0b1d043c9dce8c6517b5f5ee669e37a2b673e5b2263a6968f25364c942b351
- https://github.com/readthedocs/readthedocs.org/pull/13011/changes#diff-4eae6664ca2124d124dbfe04a8000483c7a23d426c3f390721d6e80fe5891064

I like that pattern. It uses the current workflow, but it short-circuit it to just unzip the file and move forward from there.

Author: I'm talking about managing the builders and docker, we don't need any of that to unzip a file. Unzipping is an easy operation. Again, this is left for a future improvement.

Reviewer 3: Sure, but that secondary system just brings it's own, new levels of complexity. It hasn't reduced complexity, just shifted it. I think we just scrub this from the design doc if it's not for this phase of the design.

Author: That's the flow that AWS recommends, I don't think we should scrub this from the design doc. Managing builders at a lower scale is always a win for me, lambdas are way easier than spinning a builder just to unzip a file.

Reviewer 2: Lambdas are terrible in my experience. Super limited, hard to manage dependencies and impossible to debug. If we ever want to change how we run these DAUs, I would suggest ECS at that time. We are not there, tho, and I think we won't be any time soon, tho 😄

Author: We won't be installing extra deps. The only operation the lambda would be doing is extracting the files, all post-processing is still done by our app in a celery task.

Reviewer 3: I do feel we should scrub this from the design doc. Sounds like we're mostly against using Lambda and the benefits don't really outweigh adding complexity of another AWS service to manage. Our build system works fine for our use case.

- Support OIDC for authentication instead of API tokens.
- Make latest/stable real aliases instead of copying the artifacts to a new location.
- Sign the script used to upload the artifacts.
- Do indexing tasks (search, file tree diff) just after the artifacts are uploaded to their final destination,
  instead of querying the files again from storage.

References
----------

- https://github.com/readthedocs/readthedocs.org/issues/1083
- https://github.com/readthedocs/readthedocs.org/pull/13011

Attachments

> -----------

Reviewer 2: We are testing ephemeral builds next deploy -- which will help a lot regarding security here. If we end up using with ephemeral builds + build script (`readthedocs-builder` repository), security concern will be reduced a lot.

Author: We don't really need the build script here. We no longer run arbitrary user code, having an ephemeral place to unzip things is enough.

Build anywhere, host with the best
==================================

Goals
-----

- Build the documentation anywhere, and upload the artifacts to Read the Docs.
  Zip it, ship it, host it.
- Integrate with existing modeling as much as possible.
- Easy to transition/test from our current build system for existing projects.

Non-goals
---------

- Complete re-design of the build/version system/modeling.

> - Deprecate our current build system for existing projects.

Reviewer 3: ```suggestion
- Deprecate our current build system
```

  We want to support both for the long term.

Motivation
----------

- Most requested feature since ever.

> - Security: we would no longer run arbitrary code on our servers.

Reviewer 3: ```suggestion
- Security: we would run less arbitrary code on our servers.
```

Author: what arbitrary code would we run? Unzipping isn't a user-given command.

Reviewer 3: All the arbitrary code from projects using our build system, not uploading. We're not getting away from our build system, only providing an alternative for some versions or projects.

Author: I can clarify this only applies to the new system.

This will allow us to do some checks and validations.

Versions marked as active and where their latest build is from an external upload can't be built using our current build system.
This is to avoid overwriting the documentation built from an external upload with a build from our current build system.
This should mainly happen while users migrate to the new system, new users should only use the new upload system.

We will no longer update the version to keep in sync with the Git branch/tag/commit,
as they will be created when the upload is initiated.

Versions that are deleted from the Git repository will be deleted from Read the Docs if an automation rule is configured to do so.
This includes versions created from external uploads, as they are still linked to a Git branch/tag/commit.

.. note::

   With this new system, we don't have the need to keep a copy of the Git repository versions in Read the Docs.
   We can discuss in the future how we want to handle that.
   The only reason to keep it is to be able to trigger builds from our current build system (while projects migrate),

>    and to keep automation rules working on deleted versions.

Reviewer 3: Also to clarify here, we aren't deprecating our build system, this is just a supplemental feature. We still need to handle Git repos but some projects might not need to clone the repository ever, but only if they are using all uploaded versions.

```suggestion
   With this new system, we don't have the need to keep a copy of the Git repository versions in Read the Docs for projects uploading all versions for a project.
   We still need Git repositories to be able to trigger builds using our build system
   and to keep automation rules working on deleted versions.
```

  Almost no client side validations (size, structure, etc),
  server side validations will be done when processing the build.
- Skip search configuration.
- Add the new S3 bucket with the lifecycle policy.
- Skip latest/stable logic.
- Re-use the user token for authentication.
- Minimal integration in the UI.
- Minimal checks/enforcement for versions and builds created from external uploads.

This first iteration is not a beta, it will be used for our own projects only.

Future improvements and ideas
-----------------------------

- Use a more lightweight docker image to unzip the artifacts.
- Use a lambda instead of a docker container to process the build.
  Lambdas have a maximum storage of 10GB, which may not be enough for some builds.
  But AWS has some other options for big files:

  - https://aws.amazon.com/blogs/storage/automated-extraction-of-compressed-files-on-amazon-s3-using-aws-batch-and-amazon-ecs/
  - https://aws.amazon.com/blogs/storage/automatically-decompress-files-in-amazon-s3-using-aws-step-functions/

- Support OIDC for authentication instead of API tokens.
- Make latest/stable real aliases instead of copying the artifacts to a new location.

> - Sign the script used to upload the artifacts.

Reviewer 3: I think we're really underselling the work needed in the script. Signing is nice but we really need a lot more here too.

We definitely don't want to just give a script that users copy paste into their systems or builds. This leads to having hard to upgrade scripts lying around on user systems that we have to support with backwards compatibility.

Ultimately, we want in a script:

- Easily upgraded package
- Has more complexity for validation and checks. Does uploads
- Might be pinned to a specific API schema or whatever so that we can alter the API without breaking builds

Author: The script would be distributed by us, we don't and shouldn't provide a copy/paste script.

Reviewer 3: I'm saying _we_ should not distribute anything -- not copy pasting code and not telling users to run a curl command. These are effectively the same thing. Installation should all be the job of a package manager ecosystem, and probably PyPI.

Author: Same discussion happening in https://github.com/readthedocs/readthedocs.org/pull/13134#discussion_r3478000254

- Do indexing tasks (search, file tree diff) just after the artifacts are uploaded to their final destination,
  instead of querying the files again from storage.

References
----------

- https://github.com/readthedocs/readthedocs.org/issues/1083
- https://github.com/readthedocs/readthedocs.org/pull/13011

Attachments
-----------

Artifacts size
~~~~~~~~~~~~~~

We are logging this information in
https://github.com/readthedocs/readthedocs.org/blob/c026c537983e83bb71ed489832acc1bffdee65ed/readthedocs/projects/tasks/builds.py#L1050-L1065

Using the following in New Relic

::
   SELECT count(*) FROM Log
   WHERE allColumnSearch('Build artifacts directory size.', insensitive: true)
   FACET size

We get:

::
   Size (in megabytes), Count (in thousands)
   1, 69
   10, 58
   11, 32
   13, 32
   14, 30
   2, 32
   4, 31
   5, 44
   7, 78
   8, 48

Allowing uploads of 5GB uncompressed (and 1G compressed) should be enough for most projects.

Zip file validation
~~~~~~~~~~~~~~~~~~~

> Zip file validation using unzip/zipinfo provided by Claude:

Reviewer 3: Is there a reason we're using zip file and not tar file?

Author: I'm fine with anything as long as it's a very common out of the box available command. Feels like zip is.

Reviewer 3: The two are fairly comparable formats. We might have more familiarity with tar I'm guessing. Zip is fine to use I'm guessing

I think library vs binary support is another big point for making our script in Python and not bash. We can expect the Python stdlib to have what we need where a bash script is going to rely on what is available on the system.

I'm pretty convinced we shouldn't be writing a bash script at this point.

Reviewer: I find zip easier to work with historically, and I think it's more standard for folks. Tar is very unix coded, but Mac & Windows people won't be comfortable with it.

Reviewer 2: > I'm pretty convinced we shouldn't be writing a bash script at this point.

I think we should leave the distribution/script discussion out of the scope of this PR. We should build the endpoints and infrastructure for the workflow we want to use first and test all the process by using a few cURL commands.

Once we are there, we can discuss more deeply _how_ we want to write and distribute the script --but we need the feature working first to do that.

Reviewer 3: > Tar is very unix coded, but Mac & Windows people won't be comfortable with it.

As I understood the plan, the user is going to call our script manually/in a GHA and the script is going to be responsible for creating the archive. So we can use what we feel is easiest to work with, it doesn't seem like we should need to design for users manually rolling their own as the default case.

I don't think there much tying us to either tool though.

> I think we should leave the distribution/script discussion out of the scope of this PR.

I think I agree here, if you're saying we shouldn't include hard plans for the  script as part of this design doc. It currently is described as a bash script in the design doc and conversations above though, so we need to clarify what we're aiming for here.

I don't think we should assume a bash script is a good design from this far out and agree we can build the rest of the implementation giving users a manual option before investing any time on a script. I think we don't benefit from rushing a design here.

This token should still be kept secret, as it can be used to upload builds to the pull request previews domain of the project,
but the impact of the token being leaked is much smaller, as it doesn't affect production documentation.

Webhooks and status API
-----------------------

Webhook events:

- Build triggered: we can emit this one when the upload is initiated,
  or replace it with "uploading".
- Build passed: we can emit this one when the build is finished and successful.
  Same as our current "build passed" event.
- Build failed: we can emit this one when the build is finished and failed.
  Same as our current "build failed" event.

Commit status doesn't require any changes, we can update the status as we do now with the three states: pending, success, failure.

First iteration
---------------

The first iteration would be like a half polished proof of concept, to validate the architecture and the implementation.
While not all things discussed in this document will be implemented in the first iteration,
this first iteration can be improved in the following weeks without major changes.

- Implement the upload API and the processing task.

> - Internal bash script to upload the artifacts using the API.

Reviewer 3: From all of the conversation above:

```suggestion
- Upload the artifacts using `curl` for initial iteration.
```

Author: This is very specific, but I ended up using curl and jq mainly.

Author: Added a note that this is a very basic script.
